README.md
# Identify COVID-19 Hidden Spreaders over Contact Tracing Networks

This python3 code is a demo for the paper **Identify COVID-19 Hidden Spreaders over Contact Tracing Networks**. In this demo, a spreading of COVID-19 can be simulated on an real-world network "[INFECTIOUS: STAY AWAY](http://konect.cc/networks/sociopatterns-infectious/)". Based on the symptomatic infections, the infection risk of other nodes can be ranked using the proposed method.

## Install required packages

```
pip install -r requirements.txt
```

It takes minutes to install.

## Simulate COVID-19 spreading

```
python generate.py
```

The spreaing parameters can be changed in `model.py`.

## Rank potential infections

```
python rank.py
```

It takes minutes to process. The result can be shown in `out.txt` where the first column is the index of nodes, the second column is the theoretical possibility of infection, the third column is a bool value to show if the node is infected in the simulation. In addition, a visualization of the network shows the risk of infection from low-risk nodes (in green) to high-risk nodes (in red).


spreading_data/history
100	-1	-1	-1
101	-1	-1	-1
102	-1	-1	-1
103	-1	-1	-1
104	6	-1	7
105	1	-1	2
106	0	7	-1
108	3	-1	0
109	8	-1	8
110	-1	-1	-1
111	-1	-1	-1
116	-1	-1	-1
120	-1	-1	-1
122	7	-1	7
133	-1	-1	-1
134	-1	-1	-1
10	-1	-1	-1
11	-1	-1	-1
113	-1	-1	-1
12	-1	-1	-1
13	-1	-1	-1
14	-1	-1	-1
15	-1	-1	-1
16	-1	-1	-1
107	-1	-1	-1
23	-1	-1	-1
26	-1	-1	-1
123	-1	-1	-1
128	-1	-1	-1
145	-1	-1	-1
147	-1	-1	-1
136	-1	-1	-1
124	-1	-1	-1
131	-1	-1	-1
140	-1	-1	-1
112	9	-1	2
114	-1	-1	-1
115	-1	-1	-1
125	-1	-1	-1
129	-1	-1	-1
130	-1	-1	-1
132	-1	-1	-1
167	-1	-1	-1
168	-1	-1	-1
171	-1	-1	-1
172	-1	-1	-1
177	10	-1	7
187	-1	-1	-1
188	-1	-1	-1
190	-1	-1	-1
169	-1	-1	-1
175	6	-1	6
32	-1	-1	-1
146	10	-1	10
48	-1	-1	-1
127	-1	-1	-1
183	10	-1	4
121	-1	-1	-1
138	-1	-1	-1
117	-1	-1	-1
118	-1	-1	-1
126	-1	-1	-1
148	5	-1	2
164	-1	-1	-1
178	-1	-1	-1
182	-1	-1	-1
214	-1	-1	-1
226	-1	-1	-1
233	-1	-1	-1
144	-1	-1	-1
141	-1	-1	-1
163	-1	-1	-1
166	-1	-1	-1
1	-1	-1	-1
2	-1	-1	-1
142	-1	-1	-1
17	-1	-1	-1
20	-1	-1	-1
156	8	-1	5
161	2	-1	0
160	8	-1	2
36	-1	-1	-1
149	7	-1	4
215	-1	-1	-1
143	-1	-1	-1
18	-1	-1	-1
135	-1	-1	-1
189	-1	-1	-1
222	-1	-1	-1
228	-1	-1	-1
230	7	-1	8
246	-1	-1	-1
35	-1	-1	-1
208	-1	-1	-1
49	-1	-1	-1
137	-1	-1	-1
139	-1	-1	-1
19	-1	-1	-1
165	6	-1	3
192	-1	-1	-1
218	-1	-1	-1
225	-1	-1	-1
235	-1	-1	-1
150	-1	-1	-1
151	5	-1	2
152	3	-1	3
153	3	-1	3
154	-1	-1	-1
155	-1	-1	-1
157	7	-1	4
158	-1	-1	-1
159	6	-1	6
162	9	-1	6
170	-1	-1	-1
173	-1	-1	-1
174	-1	-1	-1
179	-1	-1	-1
181	-1	-1	-1
219	-1	-1	-1
50	-1	-1	-1
51	-1	-1	-1
5	-1	-1	-1
204	-1	-1	-1
193	-1	-1	-1
224	-1	-1	-1
199	-1	-1	-1
6	-1	-1	-1
212	1	10	2
176	-1	-1	-1
191	-1	-1	-1
194	-1	-1	-1
198	-1	-1	-1
202	-1	-1	-1
206	8	-1	1
244	-1	-1	-1
247	-1	-1	-1
33	-1	-1	-1
184	-1	-1	-1
207	-1	-1	-1
255	-1	-1	-1
180	-1	-1	-1
185	-1	-1	-1
223	-1	-1	-1
238	-1	-1	-1
203	-1	-1	-1
232	-1	-1	-1
251	-1	-1	-1
253	-1	-1	-1
254	-1	-1	-1
200	-1	-1	-1
211	0	8	-1
186	-1	-1	-1
243	7	-1	5
245	-1	-1	-1
259	-1	-1	-1
264	-1	-1	-1
265	-1	-1	-1
270	0	-1	-1
271	-1	-1	-1
272	-1	-1	-1
283	-1	-1	-1
284	-1	-1	-1
286	-1	-1	-1
287	-1	-1	-1
298	-1	-1	-1
300	-1	-1	-1
308	-1	-1	-1
309	-1	-1	-1
242	-1	-1	-1
217	-1	-1	-1
213	8	-1	9
252	-1	-1	-1
201	-1	-1	-1
312	-1	-1	-1
38	10	-1	2
42	8	-1	6
195	7	-1	7
196	-1	-1	-1
197	9	-1	2
205	-1	-1	-1
210	2	-1	0
216	-1	-1	-1
229	-1	-1	-1
231	-1	-1	-1
236	-1	-1	-1
237	-1	-1	-1
241	-1	-1	-1
248	-1	-1	-1
257	-1	-1	-1
258	-1	-1	-1
260	10	-1	2
261	-1	-1	-1
262	-1	-1	-1
263	-1	-1	-1
266	-1	-1	-1
267	-1	-1	-1
277	-1	-1	-1
278	-1	-1	-1
291	-1	-1	-1
292	-1	-1	-1
299	-1	-1	-1
323	-1	-1	-1
209	-1	-1	-1
227	-1	-1	-1
239	-1	-1	-1
240	-1	-1	-1
74	-1	-1	-1
75	-1	-1	-1
76	-1	-1	-1
77	-1	-1	-1
250	-1	-1	-1
21	-1	-1	-1
22	0	-1	-1
220	-1	-1	-1
281	-1	-1	-1
288	-1	-1	-1
221	-1	-1	-1
249	-1	-1	-1
268	2	-1	3
234	-1	-1	-1
319	-1	-1	-1
321	-1	-1	-1
331	-1	-1	-1
343	-1	-1	-1
334	-1	-1	-1
335	1	-1	2
336	-1	-1	-1
353	-1	-1	-1
360	-1	-1	-1
361	-1	-1	-1
362	-1	-1	-1
294	-1	-1	-1
296	-1	-1	-1
313	-1	-1	-1
314	-1	-1	-1
39	-1	-1	-1
40	-1	-1	-1
24	-1	-1	-1
25	-1	-1	-1
47	-1	-1	-1
71	-1	-1	-1
73	-1	-1	-1
256	-1	-1	-1
302	-1	-1	-1
30	-1	-1	-1
282	-1	-1	-1
295	-1	-1	-1
297	-1	-1	-1
285	-1	-1	-1
293	-1	-1	-1
290	-1	-1	-1
315	-1	-1	-1
301	-1	-1	-1
305	-1	-1	-1
307	-1	-1	-1
274	-1	-1	-1
326	-1	-1	-1
327	-1	-1	-1
269	-1	-1	-1
70	-1	-1	-1
72	-1	-1	-1
306	-1	-1	-1
318	-1	-1	-1
324	-1	-1	-1
328	-1	-1	-1
333	-1	-1	-1
338	10	-1	8
339	-1	-1	-1
340	-1	-1	-1
341	-1	-1	-1
345	-1	-1	-1
346	-1	-1	-1
347	-1	-1	-1
348	-1	-1	-1
351	-1	-1	-1
354	-1	-1	-1
355	-1	-1	-1
356	-1	-1	-1
357	-1	-1	-1
358	-1	-1	-1
359	-1	-1	-1
365	-1	-1	-1
369	-1	-1	-1
371	-1	-1	-1
373	8	-1	9
376	-1	-1	-1
377	-1	-1	-1
379	-1	-1	-1
385	-1	-1	-1
27	-1	-1	-1
28	-1	-1	-1
29	-1	-1	-1
273	-1	-1	-1
275	-1	-1	-1
41	-1	-1	-1
44	-1	-1	-1
276	-1	-1	-1
303	-1	-1	-1
279	-1	-1	-1
280	-1	-1	-1
317	-1	-1	-1
52	-1	-1	-1
304	-1	-1	-1
289	-1	-1	-1
311	-1	-1	-1
370	-1	-1	-1
325	-1	-1	-1
310	-1	-1	-1
320	-1	-1	-1
3	-1	-1	-1
31	-1	-1	-1
316	-1	-1	-1
329	-1	-1	-1
378	-1	-1	-1
380	-1	-1	-1
381	-1	-1	-1
382	-1	-1	-1
383	-1	-1	-1
384	-1	-1	-1
386	-1	-1	-1
389	10	-1	8
322	-1	-1	-1
330	-1	-1	-1
372	-1	-1	-1
332	-1	-1	-1
337	-1	-1	-1
366	-1	-1	-1
368	-1	-1	-1
349	-1	-1	-1
363	-1	-1	-1
350	-1	-1	-1
344	-1	-1	-1
34	-1	-1	-1
37	-1	-1	-1
342	-1	-1	-1
367	-1	-1	-1
374	-1	-1	-1
375	-1	-1	-1
387	10	-1	11
395	-1	-1	-1
399	-1	-1	-1
400	0	-1	-1
404	-1	-1	-1
63	-1	-1	-1
69	-1	-1	-1
4	-1	-1	-1
402	-1	-1	-1
352	-1	-1	-1
364	-1	-1	-1
388	-1	-1	-1
391	-1	-1	-1
394	-1	-1	-1
66	-1	-1	-1
392	-1	-1	-1
396	-1	-1	-1
406	-1	-1	-1
407	-1	-1	-1
408	-1	-1	-1
393	-1	-1	-1
403	-1	-1	-1
43	-1	-1	-1
62	-1	-1	-1
65	-1	-1	-1
8	-1	-1	-1
45	-1	-1	-1
410	-1	-1	-1
78	-1	-1	-1
390	-1	-1	-1
409	-1	-1	-1
95	-1	-1	-1
9	-1	-1	-1
58	-1	-1	-1
64	10	-1	8
397	9	-1	10
398	-1	-1	-1
405	-1	-1	-1
401	-1	-1	-1
84	-1	-1	-1
85	-1	-1	-1
54	-1	-1	-1
55	-1	-1	-1
57	-1	-1	-1
92	-1	-1	-1
56	-1	-1	-1
60	-1	-1	-1
67	7	-1	8
88	-1	-1	-1
93	2	-1	0
87	-1	-1	-1
46	-1	-1	-1
80	-1	-1	-1
81	-1	-1	-1
90	-1	-1	-1
53	-1	-1	-1
7	-1	-1	-1
59	-1	-1	-1
61	-1	-1	-1
68	-1	-1	-1
97	2	-1	3
86	-1	-1	-1
94	7	-1	0
79	-1	-1	-1
119	-1	-1	-1
82	-1	-1	-1
83	-1	-1	-1
89	-1	-1	-1
91	-1	-1	-1
96	3	-1	1
99	10	-1	10
98	-1	-1	-1


model.py
import numpy as np
import scipy.sparse as sp
import copy
import math
import scipy.stats as stats    

data = ['infectious']
zero = 5                     # Number of zero infections 
k = 13.49                    # Average degree                                                  
R0 = 3.5                     # Reproduction number
TE = 5.18                    # Mean presymptomatic period 
TI = 8.83                    # Mean symptomatic period 
TA = 20                      # Mean asymptomatic period 
p = 0.15                     # Ratio of asymptomatic patience
T=11                         # Spreading days
E_L_Time =  (1-p) * (TE+TI) + p*TA  # Mean virus-carry period
a = R0/(k*E_L_Time) *(1-p)   
d = R0/(k*E_L_Time) *p

inc_period = stats.lognorm.cdf(np.arange(1,41)/4.17,s=0.66)
inc_period[-1] = 1
remove_period = stats.norm.cdf((np.arange(1,31)-8.82)/3.88)
remove_period[-1] = 1
asym_period = stats.norm.cdf((np.arange(1,41)-20)/5)
asym_period[-1] = 1

def inc(time):
	return stats.lognorm.pdf(x/4.17,s=0.66)/4.17

def remove(time):
	return stats.norm.pdf((x-8.83)/3.88)/3.88

def asym(x):
	return stats.norm.pdf((x-20)/5)/5

def get_status(name):
	status = np.genfromtxt(r"./spreading_data/status", dtype=np.dtype(str))
	return dict(zip(status[:,0].astype(np.int32), status[:,-1]))
	
	
def get_data(name):
	edge = np.genfromtxt(r"./data/{}".format(name), dtype=np.int32)
	G = {}
	for (i,j) in edge[:,0:2]:
		if i in G:
			G[i].add(j)
		else:
			G[i] = set([j])
		if j in G:
			G[j].add(i)
		else:
			G[j] = set([i])
	return G, dict(zip(G.keys(),['S']*len(G)))

def spread(G,status,control,a,d,history,T):
	new_status = copy.deepcopy(status)
	for i in status:
		if status[i][0] == 'E' or status[i][0] == 'I' or status[i][0] == 'A':
			if status[i] != 'IR' and status[i] != 'AR':
				for j in G[i]:
					if status[j] == 'S':
						#print(j)
						r = np.random.uniform(0,1)
						if r < a:
							rr = np.random.uniform(0,1)
							rr = np.where(inc_period>=rr)[0][0]+1
							new_status[j] = 'E'+str(rr)
							if history[j][2] == -1:
								history[j][2] = T-history[i][2]
							elif history[j][2] > T-history[i][2]:
								history[j][2] = T-history[i][2]
							history[j][0] = T
						elif r>a and r<a+d:
							rr = np.random.uniform(0,1)
							rr = np.where(asym_period>=rr)[0][0]+1
							new_status[j] = 'A'+str(rr)
							if history[j][2] == -1:
								history[j][2] = T-history[i][2]
							elif history[j][2] > T-history[i][2]:
								history[j][2] = T-history[i][2]
							history[j][0] = T
				if status[i][0] == 'E':
						if T-history[i][0] == int(status[i][1:]):
							rr = np.random.uniform(0,1)
							rr = np.where(remove_period>=rr)[0][0]+1
							new_status[i] = 'I'+str(rr+int(status[i][1:]))
				elif status[i][0] == 'I':
					if control == False:
						if T-history[i][0] == int(status[i][1:]):
							new_status[i] = 'IR'
							history[i][1] = T
					elif control == True:
						new_status[i] = 'IR'
						history[i][1] = T
				elif status[i][0] == 'A':
					if T-history[i][0] == int(status[i][1:]):
						new_status[i] = 'AR'
						history[i][1] = T
	return new_status,history

def stat(status):
	infe = 0
	for i in status:
		if status[i] != 'S':
			infe += 1
	return infe
	
def statA(status):
	infe = 0
	for i in status:
		if status[i][0] == 'E' or status[i][0] == 'I' or status[i][0] == 'A':
			if status[i] != 'IR' and status[i] != 'AR':
				infe += 1
	return infe

def get_out(name,G,status):
	detect = []
	one_hot = np.identity(len(G))
	Neighbor = set()
	fileobj = open(r'./spreading_data/train', 'w')
	for i in status:
		if status[i][0] == 'I' or status[i] == 'IR':
			fileobj.write(str(i)+'\t')
			for j in one_hot[i-1]:
				fileobj.write(str(j)+'\t')
			fileobj.write(str(1)+'\n')
	fileobj.close()	


generate.py
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt
from model import get_data,spread,stat,statA,get_out,data,T,zero,p,inc_period,asym_period,a,d

G,status = get_data(data[0])

history = dict(zip(G.keys(), [[-1,-1,-1] for i in range(1,len(G)+1)]))
origin = np.random.randint(1,len(G),zero)

for i in origin:
	r = np.random.uniform(0,1)
	if r < p:
		rr = np.random.uniform(0,1)
		rr = np.where(asym_period>=rr)[0][0]+1
		status[i] = 'A'+str(rr)
	else:
		rr = np.random.uniform(0,1)
		rr = np.where(inc_period>=rr)[0][0]+1
		status[i] = 'E'+str(rr)
	history[i][0] = 0

for j in range(1,int(T)):
	status,history = spread(G,status,False,a,d,history,j)


fileobj = open('./spreading_data/history','w')
for item in history:
	fileobj.write(str(item) + '\t' +  str(history[item][0])+ '\t' +  str(history[item][1])+ '\t' +  str(history[item][2])+'\n')
fileobj.close()

fileobj = open('./spreading_data/status','w')
for i in status:
	fileobj.write(str(i) + '\t' +  str(status[i])+'\n')
fileobj.close()

get_out(data[0],G,status)


rank.py
import numpy as np
import copy
import scipy.stats as stats    
import networkx as nx
import matplotlib.pyplot as plt
from model import get_data,get_status,stat,statA,data,a,d,T as ori_T,p

beta = a+d
T = ori_T

inc_prob = 1-stats.lognorm.cdf(np.arange(0,999)/4.17,s=0.66)
remove_prob = 1-stats.norm.cdf((np.arange(0,999)-8.83)/3.88)
asym_prob = 1-stats.norm.cdf((np.arange(0,999)-20)/5)

inc_prob_ra = [inc_prob[i]/inc_prob[i-1] for i in range(1,40)]
remove_prob_ra = [remove_prob[i]/remove_prob[i-1] for i in range(1,40)]
asym_prob_ra = [asym_prob[i]/asym_prob[i-1] for i in range(1,40)]

def load_data(status):
	idx_features_labels1 = np.genfromtxt(r'./spreading_data/train', dtype=np.int32)
	try:
		idx_train = set(idx_features_labels1[:,0])
	except:
		idx_train = set([idx_features_labels1[0]])
	
	return idx_train
	
def simulate_spread(state,G,idx_train,nodes):
	
	new_state = {}
	for i in nodes:
		new_state[i] = np.empty(77)
	for i in idx_train:
		new_state[i] = state[i]
	''' predict: 0S 1-20E 21-40I 41-75A 76R'''
	for t in range(1,T):
		for i in nodes:
			new_state[i][-1] = state[i][-1] + state[i][40] + state[i][75]
			new_state[i][21] = state[i][20]
			for j in range(2,21):
				new_state[i][j] = state[i][j-1]*inc_prob_ra[j-2]
				new_state[i][21] += state[i][j-1]*(1-inc_prob_ra[j-2])
			for j in range(22,41):
				new_state[i][j] = state[i][j-1]*remove_prob_ra[j-22]
				new_state[i][-1] += state[i][j-1]*(1-remove_prob_ra[j-22])
			for j in range(42,76):
				new_state[i][j] = state[i][j-1]*asym_prob_ra[j-42]
				new_state[i][-1] += state[i][j-1]*(1-asym_prob_ra[j-42])
			
			neighbor_inf = 1
			for nei in G[i]:
				if len(state[nei]) == 2:
					if t>state[nei][0] and t <= state[nei][1]:
						neighbor_inf *= (1-beta)
				else:
					neighbor_inf *= (1-beta*(sum(state[nei][1:21])+sum(state[nei][41:76])))
			
			new_state[i][1] = state[i][0]*(1-p)*(1-neighbor_inf)
			new_state[i][41] = state[i][0]*p*(1-neighbor_inf)
			new_state[i][0] = state[i][0]*neighbor_inf

		state = copy.deepcopy(new_state)
	
	return state
	
	
G,status = get_data(data[0])
idx_train = load_data(status)

nodes = set(list(G.keys()))-idx_train

status = np.genfromtxt(r"./spreading_data/status", dtype=np.dtype(str))
status = dict(zip(status[:,0].astype(np.int32), status[:,-1]))

history = np.genfromtxt(r"./spreading_data/history", dtype=np.int32)
history = dict(zip(history[:,0],history[:,1:]))

state = {}
for i in G:
	if i in idx_train:
		if history[i][1] == -1:
			state[i] = (history[i][0],T)
		else:
			state[i] = (history[i][0],history[i][1])
	else:
		state[i] = np.zeros(77)
		state[i][0] = 1

state = simulate_spread(state,G,idx_train,nodes)


ans = np.array([ [i,sum(state[i][1:21])+sum(state[i][41:76])] for i in nodes ])

temp_ans = np.argsort(ans[:,1])

fileobj = open(r'./out.txt','w')

for i in temp_ans:
	fileobj.write(str(int(ans[i][0]))+'\t'+str(ans[i][1])+'\t'+str(int(status[ans[i][0]]!='S'))+'\n')
			
fileobj.close()

edges = np.genfromtxt("./data/"+data[0], dtype=np.int32)[:,0:2]
G=nx.Graph()
G.add_edges_from(list(edges))
pos = nx.spring_layout(G)
sym_patient = np.array([[i,np.max(ans[:,1])] for i in idx_train])
all_node_color = np.vstack((ans,sym_patient)) 
nx.draw(G, pos, node_color=all_node_color[np.argsort(all_node_color[:,0])][:,1], node_size=200, cmap=plt.cm.RdYlGn)
plt.savefig('rank.svg')
